tendermint/src/merkle.rs

```diff
@@ -61,88 +61,31 @@ pub trait MerkleHash {
             },
         }
     }
-}
-
-// A helper to copy GenericArray into the human-friendly Hash type.
-fn copy_to_hash(output: impl AsRef<[u8]>) -> Hash {
-    let mut hash_bytes = [0u8; HASH_SIZE];
-    hash_bytes.copy_from_slice(output.as_ref());
-    hash_bytes
-}
-
-impl<H> MerkleHash for H
-where
-    H: Digest<OutputSize = U32> + FixedOutputReset,
-{
-    fn empty_hash(&mut self) -> Hash {
-        // Get the output of an empty digest state.
-        let digest = self.finalize_reset();
-        copy_to_hash(digest)
-    }
-
-    fn leaf_hash(&mut self, bytes: &[u8]) -> Hash {
-        // Feed the data to the hasher, prepended with 0x00.
-        Digest::update(self, [0x00]);
-        Digest::update(self, bytes);
```

Comment:
Note how this does not allocate any memory temporarily or copy data around. This is what we want most of our users to actually use, while the few platforms where the `NonIncremental` wrapper is needed could be kept as integration freak cases.